src/app/backend/data-representation/sheet-timing.ts

```
@@ -1,31 +1,66 @@
 import { TimingEntry } from './timing-entry';
 
+/**
+ * Immutable.
+ *
+ * This class contains all timing data for a sheet in a single timing file.
+ */
 export class SheetTiming {
 
     /**
-     * ID of this timing entry
+     * Id of the timing file to which this belongs
      */
-    private timingId: number;
+    private readonly timingFileId: number;
 
     /**
-     * Time the sheet should start at
+     * Time the sheet-state should start at
      */
-    private scheduledMoment: number;
+    private readonly scheduledMoment: number;
 
     /**
      * List of all timing entries
      */
-    private timingEntries: TimingEntry[] = [];
+    private readonly timingEntries: TimingEntry[];
 
-    constructor(timingId: number, scheduledMoment: number) {
-        this.timingId = timingId;
+    /**
+     * Constructs a new {@code sheetTiming}. No parameter can be undefined.
+     *
+     * @param timingId Id of the timing file to which this belongs
+     * @param scheduledMoment Time at which this sheet is scheduled to enter the system, in the associated timing file
+     * @param timingEntries Array of {@code TimingEntries} as defined in the timing file
+     */
+    constructor(timingId: number, scheduledMoment: number, timingEntries: TimingEntry[]) {
+        this.timingFileId = timingId;
         this.scheduledMoment = scheduledMoment;
+        this.timingEntries = timingEntries;
+    }
+
+    /**
+     * Finds the last {@code TimingEntry} before or equal to {@code time}.
+     * In case there is none, the last {@code TimingEntry} is returned.
+     * This assumes that all the timing entries in the data are ordered on time.
+     */
+    getLastTimingEntryAt(time: number) {
+
+        if (this.timingEntries.length === 0 || time < this.timingEntries[0].getTime()) {
+            return undefined;
```

I would not return undefined here, since this will not be checked by the engine. Instead, I would throw a parsing error if there are 0 timing entries (i.e. no timing file is selected, which should not happen at all). 

Secondly, the second check `time < this.timingEntries[0].getTime()` I would say that it throws an `IllegalArgumentError` since this should not be possible by the engine

---

Fixed